File: callback.js
const foo = (callback, arg1, arg2, arg3) => {
    console.log("FOO!", arg1, arg2, arg3);
    setTimeout(callback, 0, arg1, arg2, arg3);
};

const bar = (arg1, arg2, arg3) => {
    console.log("BAR!", arg1, arg2, arg3);
};


function main() {
    console.log("HERE!");
    setTimeout(foo, 0, bar, 1, 2, 3);
    setTimeout(foo, 0, bar, 4, 5, 6);
    console.log("THERE!");
}

main();

/*
HERE!
THERE!
FOO! 1 2 3
FOO! 4 5 6
BAR! 1 2 3
BAR! 4 5 6
*/

File: callback.py
import asyncio


def foo(loop, callback, *args):
    print("FOO!", args)
    loop.call_soon(callback, *args)


def bar(*args):
    print("BAR!", args)


def finish(loop):
    print("FINISH ME!")
    loop.stop()


def raise_exception(*args):
    raise Exception(args)


def chain(loop, *args):
    print("Chain length:", len(args))
    func, *new_args = args
    loop.call_soon(func, loop, *new_args)


def main():
    loop = asyncio.new_event_loop()
    loop.call_later(1, finish, loop)
    loop.call_soon(foo, loop, bar, 1, 2, 3)
    loop.call_soon(foo, loop, bar, 4, 5, 6)
    print("HERE")
    loop.run_forever()
    print("THERE")

    loop.call_soon(chain, loop, chain, chain, raise_exception, chain)
    loop.call_later(1, finish, loop)
    print("HERE")
    loop.run_forever()
    print("THERE")


main()
